multi_import/importer.py

```diff
@@ -0,0 +1,252 @@
+        self.model = model
+        self.lookup_fields = lookup_fields
+        self.queryset = queryset
+        self.serializer = serializer
+        self.cached_query = self.get_cached_query()
+
+    def get_serializer_context(self, context=None):
+        if context is None:
+            context = {}
+        context['cached_query'] = self.cached_query
+        return context
+
+    def can_update_object(self, instance):
+        return True
+
+    def get_cached_query(self):
+        return CachedQuery(self.queryset, self.lookup_fields)
+
+    def lookup_model_object(self, row):
+        return self.cached_query.lookup(self.lookup_fields, row.data)
+
+    def get_result_object(self, input):
+        return ImportResult(self.key, self.model)
+
+    def enumerate_data(self, data):
+        """
+        Enumerates rows, and calculates row and line numbers
+        """
+        first_row_line_number = 2
+        line_count = 0
+
+        for row_number, row_data in enumerate(data, start=1):
+            line_number = first_row_line_number + line_count
+            line_count += 1 + sum([
+                value.count('\n') for value in row_data.values()
+            ])
+            yield Row(row_number, line_number, row_data)
+
+    def transform_input(self, input_data):
+        pass
+
+    def process_changes(self, result, row, serializer):
+        pass
+
+    def run(self, data, context=None):
+        import_behaviour = self.import_behaviour(self.key,
+                                                 self.model,
+                                                 self.serializer)
+
+        result = import_behaviour.get_result_object(data)
+
+        input_data = import_behaviour.transform_input(data)
+
+        for row in self.enumerate_data(input_data):
+            try:
+                instance = self.lookup_model_object(row)
+            except MultipleObjectsReturned:
+                result.add_row_error(row,
+                                     self.error_messages['multiple_matches'])
+                continue
+
+            context = self.get_serializer_context(context)
+            serializer = self.serializer(instance=instance,
+                                         data=row.data,
+                                         context=context,
+                                         partial=True)
+
+            is_valid = serializer.is_valid()
+            update_chk = instance and (serializer.has_changes or not is_valid)
+
+            if update_chk and not self.can_update_object(instance):
+                result.add_row_error(row, self.error_messages['cannot_update'])
+                continue
+
+            if not is_valid:
+                for column_name, messages in serializer.errors.iteritems():
+                    result.add_row_errors(row, messages, column_name)
+                continue
+
+            import_behaviour.process_changes(result, row, serializer)
+
+        return result
+
+
+class ImportBehaviour(object):
```

Comment:
ImportBehaviour is the best I could come up with at the time. Suggestions are welcome :)